File: src/lib/purpose.ts
// lib/purpose.ts

import { invoke } from '@tauri-apps/api/core';
import { logger } from './logger';

export interface Purpose {
  id: string;  // UUID
  label: string;
  icon_name: string;
  is_deleted: boolean;
}

export interface CreatePurposeRequest {
  label: string;
  icon_name: string;
}

export const PurposeApi = {
  async createPurpose(purpose: CreatePurposeRequest, username: string, password: string): Promise<Purpose> {
    try {
      logger.log(`Creating new purpose: ${purpose.label}`, 'info');
      const result = await invoke('create_purpose', { purpose, username, password });
      logger.log(`Successfully created purpose: ${purpose.label}`, 'success');
      return result as Purpose;
    } catch (error) {
      logger.log(`Failed to create purpose: ${error}`, 'error');
      throw error;
    }
  },

  async getAllPurposes(includeDeleted: boolean = false): Promise<Purpose[]> {
    try {
      logger.log('Fetching purposes', 'info');
      const purposes = await invoke('get_all_purposes', { includeDeleted });
      logger.log(`Successfully fetched ${(purposes as Purpose[]).length} purposes`, 'success');
      return purposes as Purpose[];
    } catch (error) {
      logger.log(`Failed to fetch purposes: ${error}`, 'error');
      throw error;
    }
  },

  async getPurpose(id: string): Promise<Purpose> {
    try {
      logger.log(`Fetching purpose with id: ${id}`, 'info');
      const result = await invoke('get_purpose', { id });
      logger.log(`Successfully fetched purpose: ${(result as Purpose).label}`, 'success');
      return result as Purpose;
    } catch (error) {
      logger.log(`Failed to fetch purpose ${id}: ${error}`, 'error');
      throw error;
    }
  },

  async getPurposeByLabel(label: string): Promise<Purpose> {
    try {
      logger.log(`Fetching purpose with label: ${label}`, 'info');
      const result = await invoke('get_purpose_by_label', { label });
      logger.log(`Successfully fetched purpose with label: ${label}`, 'success');
      return result as Purpose;
    } catch (error) {
      logger.log(`Failed to fetch purpose with label ${label}: ${error}`, 'error');
      throw error;
    }
  },

  async updatePurpose(id: string, purpose: CreatePurposeRequest, username: string, password: string): Promise<Purpose> {
    try {
      logger.log(`Updating purpose ${id}`, 'info');
      const result = await invoke('update_purpose', { id, purpose, username, password });
      logger.log(`Successfully updated purpose ${id}`, 'success');
      return result as Purpose;
    } catch (error) {
      logger.log(`Failed to update purpose ${id}: ${error}`, 'error');
      throw error;
    }
  },

  async softDeletePurpose(id: string, username: string, password: string): Promise<void> {
    try {
      logger.log(`Soft deleting purpose ${id}`, 'info');
      await invoke('soft_delete_purpose', { id, username, password });
      logger.log(`Successfully soft deleted purpose ${id}`, 'success');
    } catch (error) {
      logger.log(`Failed to soft delete purpose ${id}: ${error}`, 'error');
      throw error;
    }
  },

  async restorePurpose(id: string, username: string, password: string): Promise<void> {
    try {
      logger.log(`Restoring purpose ${id}`, 'info');
      await invoke('restore_purpose', { id, username, password });
      logger.log(`Successfully restored purpose ${id}`, 'success');
    } catch (error) {
      logger.log(`Failed to restore purpose ${id}: ${error}`, 'error');
      throw error;
    }
  }
};

File: gj7-admin/src-tauri/src/db/purpose.rs
// src/db/purpose.rs

use uuid::Uuid;
use rusqlite::{params, Connection, Result};
use serde::{Serialize, Deserialize};
use log::info;
use rusqlite::Result as SqlResult;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Purpose {
    pub id: Uuid,
    pub label: String,
    pub icon_name: String,
    pub is_deleted: bool,  // New field for soft delete
}

#[derive(Debug, Deserialize, Clone)]
pub struct CreatePurposeRequest {
    pub label: String,
    pub icon_name: String,
}

pub trait PurposeRepository {
    fn create_purpose(&self, conn: &Connection, purpose: CreatePurposeRequest) -> Result<Purpose>;
    fn get_purpose(&self, conn: &Connection, id: Uuid) -> Result<Purpose>;
    fn get_purpose_by_label(&self, conn: &Connection, label: &str) -> Result<Purpose>;
    fn update_purpose(&self, conn: &Connection, id: Uuid, purpose: CreatePurposeRequest) -> Result<Purpose>;
    fn soft_delete_purpose(&self, conn: &Connection, id: Uuid) -> Result<()>;  // New method
    fn restore_purpose(&self, conn: &Connection, id: Uuid) -> Result<()>;
    fn get_all_purposes(&self, conn: &Connection, include_deleted: bool) -> Result<Vec<Purpose>>;  // Modified
}

pub struct SqlitePurposeRepository;

impl PurposeRepository for SqlitePurposeRepository {
    fn create_purpose(&self, conn: &Connection, purpose: CreatePurposeRequest) -> Result<Purpose> {
        let id = Uuid::new_v4();

        if purpose.label.is_empty() {
            return Err(rusqlite::Error::InvalidParameterName("Purpose label cannot be empty".to_string()));
        }

        conn.execute(
            "INSERT INTO purposes (id, label, icon_name, is_deleted) VALUES (?1, ?2, ?3, ?4)",
            params![id.to_string(), purpose.label, purpose.icon_name, false],
        )?;

        let created_purpose = Purpose {
            id,
            label: purpose.label,
            icon_name: purpose.icon_name,
            is_deleted: false,
        };

        info!("Created purpose: {}", created_purpose.label);
        Ok(created_purpose)
    }

    fn get_purpose(&self, conn: &Connection, id: Uuid) -> Result<Purpose> {
        let purpose = conn.query_row(
            "SELECT id, label, icon_name, is_deleted FROM purposes WHERE id = ?1",
            params![id.to_string()],
            |row| {
                Ok(Purpose {
                    id: Uuid::parse_str(&row.get::<_, String>(0)?).unwrap(),
                    label: row.get(1)?,
                    icon_name: row.get(2)?,
                    is_deleted: row.get(3)?,
                })
            },
        )?;

        Ok(purpose)
    }

    fn get_purpose_by_label(&self, conn: &Connection, label: &str) -> Result<Purpose> {
        let purpose = conn.query_row(
            "SELECT id, label, icon_name, is_deleted FROM purposes WHERE label = ?1 AND is_deleted = FALSE",
            params![label],
            |row| {
                Ok(Purpose {
                    id: Uuid::parse_str(&row.get::<_, String>(0)?).unwrap(),
                    label: row.get(1)?,
                    icon_name: row.get(2)?,
                    is_deleted: row.get(3)?,
                })
            },
        )?;

        Ok(purpose)
    }

    fn update_purpose(&self, conn: &Connection, id: Uuid, purpose: CreatePurposeRequest) -> Result<Purpose> {
        conn.execute(
            "UPDATE purposes SET label = ?1, icon_name = ?2 WHERE id = ?3 AND is_deleted = FALSE",
            params![purpose.label, purpose.icon_name, id.to_string()],
        )?;

        self.get_purpose(conn, id)
    }

    fn restore_purpose(&self, conn: &Connection, id: Uuid) -> Result<()> {
        // Check if a purpose with the same label exists and is not deleted
        let purpose = self.get_purpose(conn, id)?;
        
        // Check if another active purpose exists with the same label
        let existing = conn.query_row(
            "SELECT id FROM purposes WHERE label = ?1 AND is_deleted = FALSE AND id != ?2",
            params![purpose.label, id.to_string()],
            |row| row.get::<_, String>(0)
        );

        match existing {
            Ok(_) => {
                return Err(rusqlite::Error::InvalidParameterName(
                    format!("An active purpose with label '{}' already exists", purpose.label)
                ));
            }
            Err(rusqlite::Error::QueryReturnedNoRows) => {
                // No conflict found, proceed with restore
                conn.execute(
                    "UPDATE purposes SET is_deleted = FALSE WHERE id = ?1",
                    params![id.to_string()],
                )?;

                info!("Restored purpose with id: {}", id);
                Ok(())
            }
            Err(e) => Err(e),
        }
    }

    fn soft_delete_purpose(&self, conn: &Connection, id: Uuid) -> Result<()> {
        conn.execute(
            "UPDATE purposes SET is_deleted = TRUE WHERE id = ?1",
            params![id.to_string()],
        )?;

        info!("Soft deleted purpose with id: {}", id);
        Ok(())
    }

    fn get_all_purposes(&self, conn: &Connection, include_deleted: bool) -> Result<Vec<Purpose>> {
        let sql = if include_deleted {
            "SELECT id, label, icon_name, is_deleted FROM purposes"
        } else {
            "SELECT id, label, icon_name, is_deleted FROM purposes WHERE is_deleted = FALSE"
        };
        
        let mut stmt = conn.prepare(sql)?;
        
        let purpose_iter = stmt.query_map([], |row| {
            Ok(Purpose {
                id: Uuid::parse_str(&row.get::<_, String>(0)?).unwrap(),
                label: row.get(1)?,
                icon_name: row.get(2)?,
                is_deleted: row.get(3)?,
            })
        })?;

        let mut purposes = Vec::new();
        for purpose in purpose_iter {
            purposes.push(purpose?);
        }

        Ok(purposes)
    }
}

pub fn create_purposes_table(conn: &Connection) -> SqlResult<()> {
    // First create the table without the filtered unique constraint
    conn.execute(
        "CREATE TABLE IF NOT EXISTS purposes (
            id TEXT PRIMARY KEY,
            label TEXT NOT NULL,
            icon_name TEXT NOT NULL,
            is_deleted BOOLEAN NOT NULL DEFAULT FALSE
        )",
        [],
    )?;

    // Then create a unique index with the WHERE clause
    conn.execute(
        "CREATE UNIQUE INDEX IF NOT EXISTS idx_purposes_label 
         ON purposes(label) 
         WHERE is_deleted = FALSE",
        [],
    )?;

    Ok(())
}